lib/ffi-icu/duration_formatting.rb

```diff
@@ -0,0 +1,282 @@
+        STYLES_TO_LIST_JOIN_FORMAT = {
+            long: :wide,
+            short: :short,
+            narrow: :narrow,
+            digital: :narrow,
+        }
+        UNIT_FORMAT_STRINGS = {
+            years: 'measure-unit/duration-year',
+            months: 'measure-unit/duration-month',
+            weeks: 'measure-unit/duration-week',
+            days: 'measure-unit/duration-day',
+            hours: 'measure-unit/duration-hour',
+            minutes: 'measure-unit/duration-minute',
+            seconds: 'measure-unit/duration-second',
+            milliseconds: 'measure-unit/duration-millisecond',
+            microseconds: 'measure-unit/duration-microsecond',
+            nanoseconds: 'measure-unit/duration-nanosecond',
+        }
+        STYLES_TO_NUMBER_FORMAT_WIDTH = {
+            long: 'unit-width-full-name',
+            short: 'unit-width-short',
+            narrow: 'unit-width-narrow',
+            # digital for hours:minutes:seconds has some special casing.
+            digital: 'unit-width-narrow',
+        }
+
+        def self.format(fields, locale:, style: :long)
+            DurationFormatter.new(locale: locale, style: style).format(fields)
+        end
+
+        class DurationFormatter
+            def initialize(locale:, style: :long)
+                if !Lib.respond_to?(:unumf_openForSkeletonAndLocale) || !Lib.respond_to?(:ulistfmt_openForType)
+                    raise "ICU::DurationFormatting requires ICU >= 67"
+                end
+
+                raise ArgumentError, "Unknown style #{style}" unless VALID_STYLES.include?(style)
+
+                @locale = locale
+                @style = style
+                # These are created lazily based on what parts are actually included
+                @number_formatters = {}
+
+                list_join_format = STYLES_TO_LIST_JOIN_FORMAT.fetch(style)
+                @list_formatter = FFI::AutoPointer.new(
+                    Lib.check_error { |error|
+                        Lib.ulistfmt_openForType(@locale, :units, list_join_format, error)
+                    },
+                    Lib.method(:ulistfmt_close)
+                )
+            end
+
+            def format(fields)
+                fields.each_key do |field|
+                    raise "Unknown field #{field}" unless VALID_FIELDS.include?(field)
+                end
+                fields = fields.dup # we might modify this argument.
+
+                # Intl.js spec says that rounding options affect only the smallest unit, and only
+                # if that unit is sub-second. All other fields therefore need to be truncated.
+                smallest_unit = VALID_FIELDS[fields.keys.map { |k| VALID_FIELDS.index(k) }.max]
+                fields.each_key do |k|
+                    raise ArgumentError, "Negative durations are not yet supported" if fields[k] < 0
+                    fields[k] = fields[k].to_i unless k == smallest_unit && ROUNDABLE_FIELDS.include?(smallest_unit)
```

Comment:
You are not consistent with your comment above. It seems DurationFormatter on Intl.js side is still under debate and they want `{ seconds: "narrow", milliseconds: "narrow", fractionalDigits: 2}`. Would you fix the comment or adopt to their [proposal](https://github.com/tc39/proposal-intl-duration-format)?

Reply:
Again, sorry I'm not quite following.

This line is responsible for implementing the truncation of the fraction part of all the duration components except the smallest (provided the smallest one is <= seconds). It isn't called out in the spec, that I could see, precicely what should happen if you try and format something like `{hours: 4.2, minutes: 1.4, seconds: 3.3}`. I had a look at the example implementations they had and this chromium one, for example, simply rounds down fractional components to the nearest integer - https://chromium-review.googlesource.com/c/v8/v8/+/2776518/5/src/objects/js-duration-format.cc#569

Is your feedback here that the intl.js spec supports specifying the style of each component separately, but I only implemented a single `style` parameter? (and didn't implement the `fractionalDigits` parameter either). If so, I simply had no need of them for my use-case.

If you think they should be added, I can look at doing that maybe next week - I don't think it will be very difficult, just need to override what's passed to `format_number` below.

Reply:
I wasn't clear. I was nitpicking your comment saying the rounding is sub-second.

> Is your feedback here that the intl.js spec supports specifying the style of each component separately, but I only implemented a single style parameter? (and didn't implement the fractionalDigits parameter either). If so, I simply had no need of them for my use-case.
That's alright. This is a good implementation!